javascript/modules/client/src/client-dao.ts

```diff
@@ -3,34 +3,41 @@ import {
   DaoConfig,
   DaoRole,
   IClientDaoBase,
-  IClientDaoSimpleVote,
-  IClientDaoWhitelist,
-  MintConfig,
-  TokenConfig,
+  IClientDaoERC20Voting, ICreateDaoERC20Voting,
+  IClientDaoWhitelistVoting, ICreateDaoWhitelistVoting,
   VotingConfig,
 } from "./internal/interfaces/dao";
-// import { exampleContractAbi, ExampleContractMethods } from "./internal/abi/dao";
+import {
+  DAOFactory__factory,
+  Registry__factory
+} from "@aragon/core-contracts-ethers";
+
+export { ICreateDaoERC20Voting, ICreateDaoWhitelistVoting }
+
+export class ClientDaoERC20Voting extends ClientCore
+  implements IClientDaoBase, IClientDaoERC20Voting {
 
-export class ClientDaoWhitelist extends ClientCore
-  implements IClientDaoBase, IClientDaoWhitelist {
   /** DAO related methods */
   dao = {
-    create: async (
-      _daoConfig: DaoConfig,
-      _tokenConfig: TokenConfig,
-      _mintConfig: MintConfig,
-      _votingConfig: VotingConfig,
-      _gsnForwarder?: string,
-    ): Promise<string> => {
-      // TODO: This is an ethers.js integration example
-      // const instance = this.attachContractExample(
-      //   "0x1234567890123456789012345678901234567890"
-      // );
-      // const tx = await instance.store("0x1234");
-      // await tx.wait();
+    create: async (params: ICreateDaoERC20Voting): Promise<string> => {
+      if(!this.signer) throw new Error("A signer is needed for creating a DAO");
+      const daoFactoryContract = DAOFactory__factory.connect(this.daoFactoryAddress, this.signer.connect(this.web3));
 
-      // TODO: Not implemented
-      return Promise.resolve("0x1234567890123456789012345678901234567890");
+      const registry = await daoFactoryContract.registry()
+        .then(registryAddress => {
+          return Registry__factory.connect(registryAddress, this.web3)
+        });
+
+      let daoAddress = ''
+      registry.on("NewDAORegistered", (dao: string) => {
+        daoAddress = dao;
```

Comment:
This looks like a potential race condition. The `daoAddress` value could be set after the value is consumed.

Reply:
I would create a promise or something like a Deferred (a promise that you can "resolve" from elsewhere:

```ts

// Promise wrapper

export class Deferred<T> {
  promise: Promise<T> = new Promise((resolve, reject) => {
    this.resolve = value => {
      this.settled = true
      resolve(value)
    }
    this.reject = reason => {
      this.settled = true
      reject(reason)
    }
  })
  settled = false
  resolve: (value: T | PromiseLike<T>) => void
  reject: (error: Error) => void
}

// somewhere
const myDeferred = new Deferred<GatewayPool>()
return myDeferred

// elsewhere (1)
await myDeferred.promise

// elsewhere (2)
newDeferred.resolve(pool)    // will unlock elsewhere (1)
```